### PR TITLE
Add test cases for optimistic locking

### PR DESCRIPTION
**Pull request by bogdanvlviv**
Add test to ensure that locking_column cannot be updated explicitly.

Add test to prevent regression on creating new object with
explicit value locking_column.

Related to #28318

<hr/>

Clean up 'Optimistic Locking' implementation

<hr/>

/cc @rafaelfranca

### CODE REVIEW

**Comment by bogdanvlviv** [comment]
please review

File: activerecord/lib/active_record/locking/optimistic.rb [hunk 1]
```diff
@@ -107,8 +107,9 @@ def _update_record(attribute_names = self.attribute_names)
             affected_rows
 
           # If something went wrong, revert the locking_column value.
-          rescue Exception
-            send(lock_col + "=", previous_lock_value.to_i)
+          rescue
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/locking/optimistic.rb]: Why changing this? I believe we want to unset the value in any exception, not just subclasses of `StandardError`

File: activerecord/lib/active_record/locking/optimistic.rb [hunk 2]
```diff
@@ -63,8 +63,8 @@ def locking_enabled? #:nodoc:
 
         def increment_lock
           lock_col = self.class.locking_column
-          previous_lock_value = send(lock_col).to_i
-          send(lock_col + "=", previous_lock_value + 1)
+          previous_lock_value = _read_attribute(lock_col)
+          __send__("#{lock_col}=", previous_lock_value + 1)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/locking/optimistic.rb]: why using `__send__` instead of `send`?

File: activerecord/lib/active_record/locking/optimistic.rb [hunk 3]
```diff
@@ -63,8 +63,8 @@ def locking_enabled? #:nodoc:
 
         def increment_lock
           lock_col = self.class.locking_column
-          previous_lock_value = send(lock_col).to_i
-          send(lock_col + "=", previous_lock_value + 1)
+          previous_lock_value = _read_attribute(lock_col)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/locking/optimistic.rb]: This is now changing the behavior since models that override the reader are not going to use that method anymore. Why are we doing that?

**Review by rafaelfranca — CHANGES_REQUESTED** [review]

File: activerecord/lib/active_record/locking/optimistic.rb (hunk 1 above)
Comment (bogdanvlviv) [inline activerecord/lib/active_record/locking/optimistic.rb]: Sorry for that

Comment (rafaelfranca) [inline activerecord/lib/active_record/locking/optimistic.rb]: Don't need to apologize. it is important to try to remove code. This show we are missing test coverage for this.

File: activerecord/lib/active_record/locking/optimistic.rb (hunk 3 above)
Comment (bogdanvlviv) [inline activerecord/lib/active_record/locking/optimistic.rb]: yes. this is internal implementation. we need to correct increment value in any way. do you think this is incorrect?

Comment (rafaelfranca) [inline activerecord/lib/active_record/locking/optimistic.rb]: I'm almost sure this will break some application. Is this change necessary?

Comment (bogdanvlviv) [inline activerecord/lib/active_record/locking/optimistic.rb]: It looks more consistent.
at least I think we need to remove redundant `.to_i` from this line, we expect that `lock_version` of type integer. http://edgeguides.rubyonrails.org/active_record_querying.html#optimistic-locking

Comment (rafaelfranca) [inline activerecord/lib/active_record/locking/optimistic.rb]: Yes. I agree with removing the `to_i` but I don't think using `read_atttribute` make it more consistent. It actually may break application and the fact that we are using `__send__` instead of `write_attribute` make it more inconsistent.

Comment (rafaelfranca) [inline activerecord/lib/active_record/locking/optimistic.rb]: Is not the `to_i` to make `nil` to `0`? Maybe we don't have test coverage for this case but it seems to be the case.

Comment (bogdanvlviv) [inline activerecord/lib/active_record/locking/optimistic.rb]: I will investigate.

Comment (bogdanvlviv) [inline activerecord/lib/active_record/locking/optimistic.rb]: We can remove `to_i` because `send(lock_col)`(reader) returns "typecast_value". Also, we have test cases with `nil`. When i was investigating i found new bug and opened new PR to fix that, see #28914

File: activerecord/lib/active_record/locking/optimistic.rb [hunk 4]
```diff
@@ -63,8 +63,8 @@ def locking_enabled? #:nodoc:
 
         def increment_lock
           lock_col = self.class.locking_column
-          previous_lock_value = send(lock_col).to_i
-          send(lock_col + "=", previous_lock_value + 1)
+          previous_lock_value = send(lock_col)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/locking/optimistic.rb]: Do we have a test that increments the lock when the value was previous `nil`?

Comment (rafaelfranca) [inline activerecord/lib/active_record/locking/optimistic.rb]: Just saw you comment 👍

Comment (bogdanvlviv) [inline activerecord/lib/active_record/locking/optimistic.rb]: yes. https://github.com/bogdanvlviv/rails/blob/4741af97218aba5ca8f655f62bcd3b9ddda7c88d/activerecord/test/cases/locking_test.rb#L247

i explained it in previous comment https://github.com/rails/rails/pull/28833#discussion_r113808471

**Comment by bogdanvlviv** [comment]
@rafaelfranca could you please review one more time?

**Comment by bogdanvlviv** [comment]
@rafaelfranca can we finish this patch?

**Comment by bogdanvlviv** [comment]
@rafaelfranca I already rebased the branch with master, would like to finish this patch.

**Comment by bogdanvlviv** [comment]
@rafaelfranca any chance on review?

**Comment by rafaelfranca** [comment]
Thank ❤️
